pkg/query/endpointset.go

```diff
@@ -345,28 +354,125 @@ func (e *EndpointSet) Update(ctx context.Context) {
 				"address", addr, "extLset", extLset, "duplicates", fmt.Sprintf("%v", stats[component.Sidecar][extLset]+stats[component.Rule][extLset]+1))
 		}
 		stats[er.ComponentType()][extLset]++
+		if _, ok := currentEndpoints[er.addr]; !ok {
+			level.Info(e.logger).Log("msg", fmt.Sprintf("adding new %v with %+v", er.ComponentType(), er.apisPresent()), "address", addr, "extLset", extLset)
+		}
+	}
 
-		endpoints[addr] = er
-		e.updateEndpointStatus(er, nil)
+	e.endpointsMetric.Update(stats)
+}
 
-		level.Info(e.logger).Log("msg", fmt.Sprintf("adding new %v with %+v", er.ComponentType(), er.apisPresent()), "address", addr, "extLset", extLset)
+func (e *EndpointSet) updateEndpoint(ctx context.Context, spec *GRPCEndpointSpec) {
+	e.endpointsMtx.RLock()
+	er, exists := e.endpoints[spec.Addr()]
+	e.endpointsMtx.RUnlock()
+
+	if !exists {
+		var err error
+		er, err = e.newEndpointRef(ctx, spec)
+		if err != nil {
+			level.Warn(e.logger).Log("msg", "update of node failed", "err", err, "address", spec.Addr())
+			return
+		}
 	}
 
-	e.endpointsMetric.Update(stats)
+	metadata, err := spec.Metadata(ctx, infopb.NewInfoClient(er.cc), storepb.NewStoreClient(er.cc))
+	if err != nil {
+		level.Warn(e.logger).Log("msg", "update of node failed", "err", errors.Wrap(err, "getting metadata"), "address", spec.Addr())
+	}
+	er.update(e.now, metadata, err)
+	if !exists {
+		e.addEndpoint(spec, er)
+	}
+}
+
+func (e *EndpointSet) pruneEndpoints(nextEndpoints map[string]*GRPCEndpointSpec) {
+	obsoleteEndpoints := make(map[string]*endpointRef)
+	for addr, er := range e.endpointRefs() {
+		if _, ok := nextEndpoints[addr]; !ok {
+			obsoleteEndpoints[addr] = er
+		}
+	}
+
+	for _, er := range obsoleteEndpoints {
+		e.removeEndpoint(er)
+	}
+}
+
+// pruneTimedOutEndpoints removes unhealthy endpoints for which the last
+// successful health check is older than the unhealthyEndpointTimeout.
+// Strict endpoints are not removed.
+func (e *EndpointSet) pruneTimedOutEndpoints() {
+	endpoints := e.endpointRefs()
```

Comment:
Hm, wouldn't it be easier if we would operate during full Update duration on local copy of endpoints like before?

Pros:
* Avoiding constant lock/unlock of `endpointRefs` which might slow down a bit readers of map (every query/API use on Querier I assume). Not a significant slowness in practice probably, though (to measure potentially).
* Simpler code and debugging flow - Update is atomic. No partial read upon update, e.g Fetch of endpoint in the middle of configuration changed we apply to endpoint file or DNS. I can't see a major problem, other than surprise when someone updates configuration/DNS with current code, but it's some unknown piece that we have test or think about  (partial reads). Plus it's not easy to test against, prone to semantic race.

Cons:
* Insignificantly less interactive updates since we have to wait for full update, but I don't see this being a huge problem.

Reply:
I thought about this as well, but we also have endpoints that are update in-place, so we can't easily have full atomicity (and we probably don't need to). 

However, I think we can definitely make additions and removals of endpoints atomic. Now that we have more precise tests, it should be safe and easy to make the change. I'll give it a try.

Reply:
Thanks, up to you - I just thought it will simplify things

Reply:
Additions and removals should be atomic now. It also made the code much more concise and simple.